Available predictions on CPU

alejandrods opened the pull request:
I have deleted the previous pull request and I have pulled a new one to solve the conflicts. I thought it would be easier and cleaner to fix the conflicts on this way.

- Update loss.py to enable prediction on CPU
- Update visualize.py to enable prediction on CPU
- Available CPU on Notebooks
- Available CPU prediction on app.py
- Available CPU prediction on app-video.py
- Update Readme.md

alexandrevicenzi commented on ImageColorizerArtisticTests.ipynb:
```diff
@@ -7,7 +7,17 @@
-    "os.environ['CUDA_VISIBLE_DEVICES']='0' "
+    "import torch\n",
+    "import warnings\n",
+    "warnings.filterwarnings('ignore')\n",
```
Why are you disabling warnings?

alejandrods replied:
I have enabled them again

alexandrevicenzi commented on VideoColorizer.ipynb:
```diff
@@ -83,9 +100,73 @@
+    {
+     "data": {
+      "text/html": [
+       "<video alt=\"test\" autoplay \n",
```
There's a video embed into the notebook! This should not be here I believe.

alejandrods replied:
This is the result of the prediction, I will delete it

alejandrods replied:
Done

alexandrevicenzi commented:
Overall LGTM, but I'll do a test run later to see if actually works as expected.

alexandrevicenzi commented:
One thing popped in my mind and I would like to hear @jantic opinions.

This change is introducing a behavior change because now is implicit what is being used.
I personally don't like this implicit approach over an explicit one.

Let's take this scenario:

I may have a dual GPU system (true for anyone with CPU integrated graphics).
If for some reason I misconfigure something in my setup, DeOldify will just assume no GPU is available and use CPU, which might not be desired.

Whoever is running DeOldify might not even notice that and think that is using GPU when a CPU is actually being used.

Things will get a little more complicated when people start to raise issues saying that the model has some misbehavior or similar because they were using CPU instead of GPU.

For this case, I would include some parameter somewhere, we only use CPU if the user really wants to and set this parameter.

alexandrevicenzi commented:
There's clearly a reason to use GPU, sorry if I went into the hype train about running DeOldify on CPU, but for me, this patch is a no go.

I applied your patch locally and tried to run as promised, but I was expecting something to actually work.

I don't have an outrageous PC, just an i7 8gen with 16 GB of ram, but this thing can't run on CPU.

First, it froze for about 10 minutes trying to load the model, CPU 100%, all 16gb gone and fan spinning like hell. Well, it managed to load the model :tada:, then I tried to actually run a colorization, again, the i7 started to cry, for a lot more time than loading the model, I actually had to hard reset my PC after 15 min of a crazy spinning fan because it was frozen and nothing was happening.

If you come to me and says that it runs on a Xeon E5 with 128 GB of ram I believe you. Would I want to advertise that DeOldify runs on a CPU like that? No really, I rather ask people to buy a GTX1080, is just cheaper and performs better probably.

We need to keep in mind that most of the people who actually run DeOldify or they either have a GPU or they will be happy running on Colab for free, not everyone has a crazy gamer setup.

Even though I really want to see DeOldify running on CPU and embedded devices as Coral/Jetson Nano, I believe we are a bit far from that yet.

Unless you point me out something that I did wrong when running your patch, the performance is not even close to reasonable.

What I want to spend time on is to get DeOldify running (not training) on less powerful GPUs, like 1050TI, which is a very common budget GPU for many (including me).

jantic commented:
So... I tried running this on my own computer that has a Threadripper (16 core/32 thread) and it ran CPU inference quite fast actually (2-3 seconds).  I was quite pleasantly surprised actually as I swear it used to take a lot longer. So if you have a beastly machine like mine then having CPU as an option is great as you can get past the GPU memory limitations (I've done this before for this purpose).

But that's a Threadripper.  That's an extreme case. @alexandrevicenzi has reported he's getting much worse performance on a more normal setup and I believe him.

Relatedly, he has made some excellent arguments as to why we should stick with being on GPU by default (even if there isn't any detected by PyTorch).  I totally agree that GPU should be considered selected by default until CPU is explicitly selected by the user. It's indeed convenient in a lot of ways to be implicit but at the same time it allows for the potential for needless confusion. I'd have it fail loudly if the default assumption of GPU being available isn't true yet GPU is selected.

I reviewed the changes and one thing we'll need changed is that loss.py shouldn't be altered like that for the cpu functionality, as loss functions are only used for training and training is only done on the gpu practically speaking. They're not actually used for inference.  My solution if somehow you're finding these getting passed around for inference would be to just pass another loss function like L1 that doesn't use GPU.  That being said I'm already seeing that being done in the code that instantiates inference instances so I'm not sure what ever motivated this change in the first place....

I'd also be reluctant to have this altered line included in app.py:
`port = 9000 `

That used to be 5000 and this code is being used in production by at least one group that I know of.

With those changes made I'll be happy with this personally.  Thanks!

alexandrevicenzi commented:
As @jantic pointed out, `app.py` and `app-video.py` might not even be changed, as these files are not into core functionality (or even part of DeOldify library) of DeOldify and not used by any of the notebooks.

The people using these files have no interest in running on CPU I believe.

I also agree with @jantic, we can merge this, but only if the user can explicitly ask for CPU, and should also be included in the README what kind of CPU/Memory requirements are needed as many won't be able to run on their "Walmart" laptop/pc.

alexandrevicenzi commented:
@jantic, @alejandrods may I ask what model you tried to run? Perhaps that's what made all the difference.

In my case I tried to run Stable model.

alejandrods commented:
Hi, thanks for your comments. :) 

First of all, I think that we don't have to suppose that every person has a GPU or every person want to run the code on GPU. Because, maybe, many people want to run the code in a CPU Instance on AWS or something like that (since it's cheaper) or they just don't have GPU. We should include both options. 

I don't understand the point: "I may have a dual GPU system (true for anyone with CPU integrated graphics).", basically for 2 reasons:
   - If you have Dual GPU system, you have to configure your code to run it in parallel, on both GPU, it's not by default. (Maybe this can be a point to improve the code).
   - CUDA, actually, doesn't support the integrated graphics of CPU, like Intel Iris Plus Graphics 640 or something like that.... Well, you can use integrated graphics with CUDA but you have to use other tools (OpenCL, ...) but they doesn't work well. Finally, to use GPU on an effective way you have to have a GPU with a computation power greater than 3. https://developer.nvidia.com/cuda-gpus

@jantic, you are right, `loss.py` is not used on inference but the class is initialized when we call `visualize.py`. Actually, if we don't insert the cpu funcionality in `loss.py` we get the next error:

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-e79761dd73f8> in <module>
----> 1 from deoldify.visualize import *
      2 plt.style.use('dark_background')

~/Documents/My Projects/DeOldify/deoldify/visualize.py in <module>
      5 from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas
      6 from .filters import IFilter, MasterFilter, ColorizerFilter
----> 7 from .generators import gen_inference_deep, gen_inference_wide
      8 from tensorboardX import SummaryWriter
      9 from scipy import misc

~/Documents/My Projects/DeOldify/deoldify/generators.py in <module>
     14       return learn
     15 
---> 16 def gen_learner_wide(data:ImageDataBunch, gen_loss=FeatureLoss(), arch=models.resnet101, nf_factor:int=2)->Learner:
     17     return unet_learner_wide(data, arch=arch, wd=1e-3, blur=True, norm_type=NormType.Spectral,
     18                         self_attention=True, y_range=(-3.,3.), loss_func=gen_loss, nf_factor=nf_factor)

~/Documents/My Projects/DeOldify/deoldify/loss.py in __init__(self, layer_wgts)
     11 
     12         # if torch.cuda.is_available():
---> 13         self.m_feat = models.vgg16_bn(True).features.cuda().eval()
     14         # else:
     15         #     self.m_feat = models.vgg16_bn(True).features

~/Documents/env/deoldify_cpu/lib/python3.7/site-packages/torch/nn/modules/module.py in cuda(self, device)
    303             Module: self
    304         """
--> 305         return self._apply(lambda t: t.cuda(device))
    306 
    307     def cpu(self):

~/Documents/env/deoldify_cpu/lib/python3.7/site-packages/torch/nn/modules/module.py in _apply(self, fn)
    200     def _apply(self, fn):
    201         for module in self.children():
--> 202             module._apply(fn)
    203 
    204         def compute_should_use_set_data(tensor, tensor_applied):

~/Documents/env/deoldify_cpu/lib/python3.7/site-packages/torch/nn/modules/module.py in _apply(self, fn)
    222                 # `with torch.no_grad():`
    223                 with torch.no_grad():
--> 224                     param_applied = fn(param)
    225                 should_use_set_data = compute_should_use_set_data(param, param_applied)
    226                 if should_use_set_data:

~/Documents/env/deoldify_cpu/lib/python3.7/site-packages/torch/nn/modules/module.py in <lambda>(t)
    303             Module: self
    304         """
--> 305         return self._apply(lambda t: t.cuda(device))
    306 
    307     def cpu(self):

~/Documents/env/deoldify_cpu/lib/python3.7/site-packages/torch/cuda/__init__.py in _lazy_init()
    190             raise RuntimeError(
    191                 "Cannot re-initialize CUDA in forked subprocess. " + msg)
--> 192         _check_driver()
    193         torch._C._cuda_init()
    194         _cudart = _load_cudart()

~/Documents/env/deoldify_cpu/lib/python3.7/site-packages/torch/cuda/__init__.py in _check_driver()
     93 def _check_driver():
     94     if not hasattr(torch._C, '_cuda_isDriverSufficient'):
---> 95         raise AssertionError("Torch not compiled with CUDA enabled")
     96     if not torch._C._cuda_isDriverSufficient():
     97         if torch._C._cuda_getDriverVersion() == 0:

AssertionError: Torch not compiled with CUDA enabled
```

I have the next MacbookPro: 

![image](https://user-images.githubusercontent.com/32141606/67369593-97e92e00-f579-11e9-908f-fdcaa684b5a5.png)

So I don't have a huge gamer computer, I have run the CPU inference on this computer and I got a good performance. I have done some benchmark on `ImageColorizer.ipynb`:

- Import `get_image_colorizer' takes 314 ms (average)
![image](https://user-images.githubusercontent.com/32141606/67376763-f667d980-f584-11e9-9470-bb484e8fdcf4.png)

- Import `get_image_colorizer' takes 2.26 s (average)
![image](https://user-images.githubusercontent.com/32141606/67376912-2fa04980-f585-11e9-8d16-7b4410624c50.png)

I have used these models:
https://www.dropbox.com/s/zkehq1uwahhbc2o/ColorizeArtistic_gen.pth?dl=0

I will reverse the changes I made in app.py and app-video.py

Thank you guys! :) 👍

alexandrevicenzi commented:
@alejandrods what we are trying to say is that CPU should not be chosen by default, it must be explicitly asked by the user to be used.

About your model, try to run on Stable model and check if you achieve the same performance, and also do a try on Video.

alejandrods commented:
@alexandrevicenzi Using Stable Model (on the same laptop, MacBook Pro):

- Import `get_image_colorizer' takes 12 s (average).
![image](https://user-images.githubusercontent.com/32141606/67429757-1d53f900-f5e1-11e9-9260-f6780217f45c.png)

- Inference takes 25 s (average)
![image](https://user-images.githubusercontent.com/32141606/67429772-23e27080-f5e1-11e9-9a0f-4d29f7173101.png)

On Video:
- Inference takes 4 minutes (average)
![image](https://user-images.githubusercontent.com/32141606/67429792-2cd34200-f5e1-11e9-866b-e2dbdfac83e8.png)


On the other hand, CPU is not chosen by default. We checked if CUDA exists, if we don't find CUDA, we use CPU. But if we detect GPU, we use GPU. So, GPU is by default.

```
if torch.cuda.is_available():
    os.environ['CUDA_VISIBLE_DEVICES']='0'
    torch.backends.cudnn.benchmark=True
```

Thank you! Let me know with your comments! :)

alexandrevicenzi commented:
@alejandrods What we have been trying to say is, if the user has no GPU the notebook should FAIL as it's now, unless the user configures notebooks to use CPU.

We don't want the notebook to assume what is going to be used (CPU or GPU), we want the user to choose it, manually.

Also, please, squash and rebase your patch, there are unrelated changes on it from some bad merge.

alejandrods commented:
Done!

alexandrevicenzi commented:
Squash and rebase please, it's hard to review this way, too many unrelated changes are appearing in git diff.

You can also give maintainers permissions to edit your PR, in this case we can fix any conflicts.

alejandrods commented:
Done! I have activated the maintainers permissions:

![image](https://user-images.githubusercontent.com/32141606/67474748-49a95d00-f655-11e9-989c-40c822b12ddd.png)

Thanks!

alejandrods commented:
Any news? Can we merge?

alexandrevicenzi commented on deoldify/filters.py:
```diff
@@ -14,12 +14,9 @@
-        pass
-
-
+    def filter(self, orig_image:PilImage, filtered_image:PilImage, render_factor:int)->PilImage:
```
Why this was changed?

alejandrods replied:
It has not sense to split the arguments if you are not using some style guide as PEP8.

alejandrods replied:
I have changed it anyway, but there is some standards about the style guide of the code, as PEP8. Actually, this line as others, don't follow thee PEP8 standard

alexandrevicenzi commented:
@alejandrods be patience. I'm waiting @jantic review as well.

loretoparisi commented:
@alejandrods amazing PR!!! Did you update the Dockerfile as well to run in CPU?

alejandrods commented:
@loretoparisi No, I didn't update it. I had updated app.py and app-video.py but @alexandrevicenzi told me that this isn't unnecessary. Tell me if you need it, I can update them :)

loretoparisi commented:
@alejandrods thank you, yes it would be great. Looking at the `Dockerfile` and `Dockerfile-api` are both: `FROM nvcr.io/nvidia/pytorch:19.04-py3` that is devised for gpu: `https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_19-09.html#rel_19-09`.

In fact running an image, like the notebook without `nvidia-docker` you will get an assertion error:

```
AssertionError: 
Found no NVIDIA driver on your system. Please check that you
have an NVIDIA GPU and installed a driver from
http://www.nvidia.com/Download/index.aspx
Exception ignored in: <bound method FeatureLoss.__del__ of FeatureLoss()>
Traceback (most recent call last):
  File "/data/deoldify/loss.py", line 43, in __del__
    self.hooks.remove()
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 537, in __getattr__
    type(self).__name__, name))
AttributeError: 'FeatureLoss' object has no attribute 'hooks'
```

alexandrevicenzi commented:
@alejandrods @loretoparisi this so-called app.py is not even part of DeOldify and is scheduled to be removed from this repository, in this case, don't touch it because we will happily throw all your changes away in the future.

@loretoparisi if you want a CPU Docker image just create your own and use DeOldify as a dependency/submodule on your project.

loretoparisi commented:
@alexandrevicenzi got it, thank you. As soon as the cpu PR is merged, I will prepare a new specific Dockerfile.

alejandrods commented:
- I have updated my branch with the last changes from @alexandrevicenzi, I have discarded the changes in `loss.py`. 
- I have built a function for setting cpu/gpu in a folder called `utils`. 
- I have refactored the implicit switch and it's explicit now (based on the cpu/gpu function)

Travis failed, could you tell me why? Or tell me what checks Travis does?

alexandrevicenzi commented on deoldify/filters.py:
```diff
@@ -10,16 +10,16 @@
 from scipy import misc
 import cv2
 from PIL import Image as PilImage
+from utils.device import *
```
Don't ever use `*` as import.

alejandrods replied:
Perfect!

alexandrevicenzi commented on utils/device.py:
```diff
@@ -0,0 +1,24 @@
+import torch
```
All python code should be under `DeOldify`, in this case `from deoldify import utils`.

alexandrevicenzi commented on utils/device.py:
```diff
@@ -0,0 +1,24 @@
+    else:
+        raise Exception(f'Set device = gpu or device = cpu')
+
+    device.use_gpu = use_gpu
```
I prefer to keep torch.device instance stored here, might be useful.

alejandrods commented:
@alexandrevicenzi Can you make this change, please? Actually, I'm not sure how to do it. When it's ready I would change the other changes requested.

alexandrevicenzi commented:
@alejandrods I'm going to close this and send a new PR.

alejandrods commented:
I don't understand why are you going to close the pull request? Why don't modify my pull request? If you are going to make a new one, my work is lost. I don't understand why are you proceeding on this way.
